tests/functionality/scopflow/selfcheck.cpp

```diff
@@ -104,23 +108,81 @@ struct ScopflowFunctionalityTestParameters {
+  MPI_Comm comm;
+  int nprocs;
+  int logging_rank = 0;
+
+  ScopflowFunctionalityTests(std::string testsuite_filename, MPI_Comm comm,
+                             int logging_verbosity = EXAGO_LOG_INFO)
+      : FunctionalityTestContext(testsuite_filename, logging_verbosity),
+        comm{comm} {
+    int my_rank;
+    auto rerr = MPI_Comm_rank(comm, &my_rank);
+    if (rerr)
+      throw ExaGOError("Error getting MPI rank number");
+
+    auto err = MPI_Comm_size(comm, &nprocs);
+    if (err) {
+      if (my_rank == logging_rank)
+        throw ExaGOError("Error getting MPI num ranks");
+      exit(0);
+    }
+  }
+
   void
   ensure_options_are_consistent(toml::value testcase,
                                 toml::value presets = toml::value{}) override {
+    int my_rank;
+    auto err = MPI_Comm_rank(comm, &my_rank);
+    if (err)
+      throw ExaGOError("Error getting MPI rank number");
+
+#if 0
+    int n_preset_procs;
+    set_if_found(n_preset_procs, presets, "n_procs");
+    int n_testcase_procs = -1;
+    set_if_found(n_testcase_procs, testcase, "n_procs");
+
+    if (-1 != n_testcase_procs) {
+      if (my_rank == logging_rank) {
+        std::stringstream errs;
+        errs << "Number of processes should be declared globally in the preset "
+                "area of the test suite TOML file, not inside each testcase.\n"
+             << "Testcase: " << testcase << "\nWith presets:\n"
+             << presets;
+        throw ExaGOError(errs.str().c_str());
+      }
+      exit(0);
+    } else if (nprocs != n_preset_procs) {
+      if (my_rank == logging_rank) {
+        std::stringstream errs;
+        errs << "SCOPFLOW Functionality test suite found " << n_preset_procs
+             << " processes specified in the presets of the test suite TOML "
+                "file, but this test is being run with "
+             << nprocs << " processes.\nTestcase: " << testcase
+             << "\nWith presets:\n"
+             << presets;
+        throw ExaGOError(errs.str().c_str());
+      }
+      exit(0);
+      return;
+    }
+#endif
+
     auto ensure_option_available = [&](const std::string &opt) {
       bool is_available = testcase.contains(opt) || presets.contains(opt);
       if (!is_available) {
-        std::stringstream errs;
-        errs << "SCOPFLOW Test suite expected option '" << opt
-             << "' to be available, but it was not found in this testsuite"
-             << " configuration:\n";
-        errs << testcase << "\nwith these presets:\n" << presets;
-        throw ExaGOError(errs.str().c_str());
+        if (my_rank == logging_rank) {
+          std::stringstream errs;
+          errs << "SCOPFLOW Test suite expected option '" << opt
+               << "' to be available, but it was not found in this testsuite"
+               << " configuration:\n";
+          errs << testcase << "\nwith these presets:\n" << presets;
+          throw ExaGOError(errs.str().c_str());
+        }
+        exit(0);
```

Comment:
Again void functions don't need exit? I think all ranks perhaps then should throw the error, but only one rank should print anything?

Reply:
exit stops the entire program.

Reply:
Right so maybe we want to use ExaGO's exit/error functionality to ensure a graceful cleanup

Reply:
I got rid of the exit(0)'s completely. According to Bill, the C++ exceptions should stop the program unless they are caught in a try/catch block by the calling program, so that is probably approximately the behavior we want. If it isn't, the place to fix it would be the ExaGOError functionality.